compare_platforms/compare_platforms_read_only.c

```diff
@@ -0,0 +1,147 @@
+ * compare the two.
+ * 
+ * In a CHERI environment, it will:
+ *     Step 1: Use can_read to check if the struct's "permissions" member includes write permissions.
+ *     Step 2: Use capabilities to enforce read-only permissions
+ *             (see https://github.com/capablevms/cheri-examples/blob/master/employee/include/employee.h).
+ *     Step 3: Try to write to the read-only struct members, and SIGPROT.
+ *
+ * In a non-CHERI environment, it will:
+ *     Step 1: Use can_read to check if the struct's "permissions" member includes write permissions.
+ *     Step 2: Try to exceed the bounds of an object and add write permissions
+ *             (see https://ctsrd-cheri.github.io/cheri-exercises/exercises/buffer-overflow-heap/index.html).
+ *     Step 3: Rewrite the review.
+ *
+ * In a CHERI environment, the overflow attack will be unreachable.
+ * In morello-purecap, it would crash immediately whether or not the struct
+ * had been set as read only.
+ *
+ * In a non-CHERI environment, this attack should succeed in corrupting some of the data.
+ *
+ */
+
+#include "include/review.h"
+#include <assert.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+int main()
+{
+    char *username, *realname, *privatereview, *publicreview, *permissions;
+    
+    // Our new review, which will have read only permissions.
+    struct review review;
+    
+    size_t smallsz = 0x20;
+    size_t biggersz = 0x1001;
+    
+    // Populate the members of the review struct.
+    username = malloc(smallsz);
+    strcpy(username, "Reviewer 2");    
+    size_t szn = actual_input_length(username, smallsz);
+    
+    // Reviewer's real name
+    realname = malloc(smallsz);
+    strcpy(realname, "Baba Yaga");
+    size_t szp = actual_input_length(realname, smallsz);
+    
+    // The reviewer's private comments
+    privatereview = malloc(biggersz);
+    strcpy(privatereview, "I cannot believe I read this appalling piece of dreck from start to finish. The authors should be ashamed, and I hope I get an opportunity to tell them so to their faces.");
+    size_t szprr = actual_input_length(privatereview, biggersz);
+
+    // The public review
+    publicreview = malloc(biggersz);
+    strcpy(publicreview, "I am a little unclear as to the contribution. I think the authors could strengthen their case considerably if they conducted an RCT. Weak reject.");
+    size_t szpur = actual_input_length(publicreview, biggersz);
+    
+    // The permissions
+    permissions = malloc(smallsz);
+    strcpy(permissions, "r");
+    size_t szpm = actual_input_length(permissions, smallsz);
+    
+    review.username = username;
+    review.realname = realname;
+    review.privatereview = privatereview;
+    review.publicreview = publicreview;
+    review.permissions = permissions;
+    print_details(&review);
+
+#ifdef __CHERI_PURE_CAPABILITY__
+    printf("\nsmallsz=%zx, CRRL(smallsz)=%zx\n", smallsz,
```

Comment:
What's `CRRL` here?

Reply:
"CRoundRepresentableLength(CRRL) accepts a proposed bounds length, and returns a rounded-up size that will be accepted by CSetBoundsExact without throwing an exception." UCAM-CL-TR-951, section 3.5.5.

Reply:
Do we expect the CRRL to be the same as the values we set, or not? And if not, will this affect functionality?

Reply:
With the smaller allocations, we do. With larger allocations, the CRRL might be wider. Length <= 4096 is supposed to have 1-byte alignment. Supposedly this is something for the system components to take care of, not the developer.

Reply:
I'm personally unsure if this is needed information to be printed, but I have no strong feelings. If you think it's fine as is, just resolve this thread. Otherwise, change as you see fit and update.

Reply:
I think it's fine as is.